### PR TITLE
Extend > Resources

### PR DESCRIPTION
**Pull request by appilon**
Adds a page in the sidebar after `writing custom providers` and `schemas`, covers the advanced features of Resources.

Preview:
http://tf-alex-preview2.s3-website-us-west-2.amazonaws.com/docs/extend/resources.html

I will squash the commits before merging

### CODE REVIEW

File: content/source/docs/extend/resources.html.md [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        if resp.Status != "CREATED" {
+            return resource.RetryableError(fmt.Errorf("Expected instance to be created but was in state %s", resp.Status))
+        }
+
+        return resource.NonRetryableError(resourceExampleInstanceRead(d, meta))
+    })
+}
+```
+
+### StateChangeConf
```
Comment (radeksimko) [inline content/source/docs/extend/resources.html.md]: I think the header should be something like `Retry Helpers`?

File: content/source/docs/extend/resources.html.md [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+`resource.Retry` is useful for simple scenarios, but sometimes handling an APIs latency or eventual consistency requires more fine tuning. `resource.Retry` is in fact a wrapper for a another helper: `resource.StateChangeConf`. Use `resource.StateChangeConf` when your resource has several pending states, the retry timing and backoff requires fine-tuning, or you want to ensure a minimum number of occurrences of a target state is reached (this is very common when dealing with eventually consistent APIs, where a response can reply back with an old state between calls before becoming consistent).
```
Comment (radeksimko) [inline content/source/docs/extend/resources.html.md]: I think we should explain the main difference between the two:

`resource.Retry` is useful for retrying on errors or more generally when dealing with **two** states (error / success).
`resource.StateChangeConf` should be preferred when the resource can go through **multiple** states as it provides more fine grained control and set expectations around what state can the resource be in.

It doesn't have to be word by word, but this is how I understand the two differ, and how I would decide which one to use.

File: content/source/docs/extend/resources.html.md [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    }
+
+    d.Set("name", obj.Name)
+    return nil
+}
+```
+
+There are other cases where the `READ` function uses configuration parameters as the identifier or support for importing multiple resources, as seen in the [AWS Provider](https://github.com/terraform-providers/terraform-provider-aws/blob/d3fe7e9907263b1aa41ddc0736a34b42899d1536/aws/import_aws_dx_gateway.go#L12) is needed. In general it is advised to stick to the passthrough importer when possible.
+
+## Customizing Differences
```
Comment (radeksimko) [inline content/source/docs/extend/resources.html.md]: Not necessarily a blocker, but are we going to leave out `DiffSuppressFunc` here? I feel like it's even more commonly used than `CustomizeDiff` and probably deserves a few sentences?

Comment (appilon) [reply content/source/docs/extend/resources.html.md]: DiffSuppressFunc is covered in schemas:
https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#diffsuppressfunc

This doc focuses on Resource APIs, as far as I know DiffSuppressFunc is at the schema level?

File: content/source/docs/extend/resources.html.md (hunk 2 above)
Comment (appilon) [reply content/source/docs/extend/resources.html.md]: 👍

File: content/source/docs/extend/resources.html.md (hunk 3 above)
Comment (radeksimko) [reply content/source/docs/extend/resources.html.md]: Ah, doh, you're right, I missed that.

File: content/source/docs/extend/resources.html.md (hunk 1 above)
Comment (radeksimko) [reply content/source/docs/extend/resources.html.md]: The higher-level one, sorry for the confusion. I think headers should use a human-friendly language - something the reader can understand and then we can dive into terms like `StateChangeConf` below.

File: content/source/docs/extend/resources.html.md [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        }
+
+        return resource.NonRetryableError(resourceExampleInstanceRead(d, meta))
+    })
+}
+```
+
+### StateChangeConf
+
+`resource.Retry` is useful for simple scenarios, particularly when the API response is either success or failure, but sometimes handling an APIs latency or eventual consistency requires more fine tuning. `resource.Retry` is in fact a wrapper for a another helper: `resource.StateChangeConf`. Use `resource.StateChangeConf` when your resource has multiple states to progress though, you require fine grained control of retry and delay timing, or you want to ensure a minimum number of occurrences of a target state is reached (this is very common when dealing with eventually consistent APIs, where a response can reply back with an old state between calls before becoming consistent).
```
Comment (radeksimko) [inline content/source/docs/extend/resources.html.md]: Just a nitpick - I'd put _Use `resource.StateChangeConf` when _ to a separate paragraph, i.e. insert two newlines.

**Review by radeksimko — APPROVED** [review]
Just one nitpick/minor suggestion, but feel free to :shipit:

File: content/source/docs/extend/resources.html.md (hunk 4 above)
Comment (appilon) [reply content/source/docs/extend/resources.html.md]: 👍

**Pull request merged**